File: apps/extension/lib/tagger.js
/**
 * Running the character tagger inside the browser.
 *
 * The session is injected rather than created here, so this file can be tested without
 * onnxruntime and so the same code works in a service worker, in a page, or in a test.
 *
 * Long text is cut into overlapping windows: the model reads a whole window but only the
 * middle of it is kept, so no character is ever decided without context on both sides. The
 * bounds are the same ones the Python side computes.
 */

import { restoreWithLabels, stripDiacritics } from "./alphabet.js";

const PAD_ID = 0;
const UNKNOWN_ID = 1;

/** Split a length into windows, with the part of each window that is actually used. */
export function windowBounds(length, window, overlap) {
  if (length <= window) {
    return [{ start: 0, end: length, commitStart: 0, commitEnd: length }];
  }

  const stride = window - overlap;
  const half = Math.floor(overlap / 2);
  const bounds = [];

  for (let start = 0; start < length; start += stride) {
    const end = Math.min(start + window, length);
    bounds.push({
      start,
      end,
      commitStart: start === 0 ? start : start + half,
      commitEnd: end === length ? end : end - half,
    });
    if (end === length) break;
  }

  return bounds;
}

export class LocalTagger {
  /**
   * @param run  async (ids: BigInt64Array, rows: number, columns: number) => Float32Array
   *             of shape [rows, columns, 2]; the model's two scores per character.
   * @param meta {characters: string[], config: {window: number, overlap: number}}
   */
  constructor(run, meta) {
    this.run = run;
    this.config = meta.config;
    this.ids = new Map();
    meta.characters.forEach((char, index) => this.ids.set(char, index + 2));
  }

  encode(text) {
    return [...text].map((char) => this.ids.get(char) ?? UNKNOWN_ID);
  }

  /** One label per character, for each text. */
  async predict(texts) {
    return (await this.predictDetailed(texts)).labels;
  }

  /**
   * Labels plus how sure the model was of each one.
   *
   * The two scores per character are turned into a probability with a softmax; the
   * confidence of a character is the probability of the label that was chosen, so it is
   * always at least one half.
   */
  async predictDetailed(texts) {
    const pieces = [];
    texts.forEach((text, index) => {
      const characters = [...text];
      for (const bound of windowBounds(characters.length, this.config.window, this.config.overlap)) {
        pieces.push({
          index,
          text: characters.slice(bound.start, bound.end).join(""),
          from: bound.commitStart - bound.start,
          to: bound.commitEnd - bound.start,
        });
      }
    });

    const labels = texts.map(() => []);
    const confidence = texts.map(() => []);
    if (pieces.length === 0) return { labels, confidence };

    const columns = Math.max(...pieces.map((piece) => [...piece.text].length), 1);
    const ids = new BigInt64Array(pieces.length * columns).fill(BigInt(PAD_ID));

    pieces.forEach((piece, row) => {
      this.encode(piece.text).forEach((id, column) => {
        ids[row * columns + column] = BigInt(id);
      });
    });

    const scores = await this.run(ids, pieces.length, columns);

    pieces.forEach((piece, row) => {
      for (let column = piece.from; column < piece.to; column += 1) {
        const at = (row * columns + column) * 2;
        const mark = 1 / (1 + Math.exp(scores[at] - scores[at + 1]));
        const label = mark > 0.5 ? 1 : 0;
        labels[piece.index].push(label);
        confidence[piece.index].push(label ? mark : 1 - mark);
      }
    });

    return { labels, confidence };
  }

  /** Restore one text. */
  async restore(text) {
    const [restored] = await this.restoreMany([text]);
    return restored;
  }

  /** Restore a batch; batching is what keeps this fast enough to feel instant. */
  async restoreMany(texts) {
    if (texts.length === 0) return [];
    const predictions = await this.predict(texts.map((text) => stripDiacritics(text)));
    return texts.map((text, index) => restoreWithLabels(text, predictions[index]));
  }
}

File: packages/core/schwa/tagger.py
"""The character tagger: decide for every character whether it carries a diacritic.

Unlike the lexicon and the context model, this one does not need to have seen a word before.
That is the point: most remaining errors are rare proper nouns that never appear in the
training text, and a word-level system can only leave those as typed.

Torch is imported lazily, so installing ``schwa`` does not drag a deep learning framework
onto machines that only want to restore text with the lexicon.
"""

from __future__ import annotations

import json
from collections import Counter
from collections.abc import Callable, Iterable, Sequence
from dataclasses import asdict, dataclass
from pathlib import Path

from schwa.alphabet import (
    LABEL_KEEP,
    LABEL_MARK,
    apply_labels,
    is_foldable,
    strip_diacritics,
    to_labels,
)

__all__ = ["CharVocabulary", "TaggerConfig", "TaggerRestorer", "restore_with_labels"]

PAD = "\x00"
UNKNOWN = "\x01"


@dataclass(frozen=True)
class TaggerConfig:
    """Everything needed to rebuild the model around a checkpoint."""

    embedding: int = 96
    hidden: int = 256
    layers: int = 2
    dropout: float = 0.2
    window: int = 256
    #: Windows overlap by this many characters so that a character near a cut is still
    #: predicted with context on both sides.
    overlap: int = 32


class CharVocabulary:
    """Maps characters to ids, with slots for padding and unseen characters."""

    def __init__(self, characters: Sequence[str]) -> None:
        self.characters = [PAD, UNKNOWN, *characters]
        self._ids = {char: index for index, char in enumerate(self.characters)}

    @classmethod
    def from_texts(cls, texts: Iterable[str], min_count: int = 20) -> CharVocabulary:
        """Collect the characters worth an embedding of their own."""
        counts: Counter[str] = Counter()
        for text in texts:
            counts.update(strip_diacritics(text))
        kept = sorted(char for char, count in counts.items() if count >= min_count)
        return cls([char for char in kept if char not in (PAD, UNKNOWN)])

    def encode(self, text: str) -> list[int]:
        """Return one id per character."""
        unknown = self._ids[UNKNOWN]
        return [self._ids.get(char, unknown) for char in text]

    def save(self, path: Path) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(self.characters[2:], ensure_ascii=False), encoding="utf-8")

    @classmethod
    def load(cls, path: Path) -> CharVocabulary:
        return cls(json.loads(path.read_text(encoding="utf-8")))

    def __len__(self) -> int:
        return len(self.characters)


def restore_with_labels(text: str, predicted: Sequence[int]) -> str:
    """Apply predicted labels to ``text``, keeping the diacritics the user typed.

    Someone who typed ``mən sence`` already told us about the first word. Those characters
    keep their own reading and only the plain ASCII ones are left to the model.
    """
    typed, written = to_labels(text)
    labels = [
        LABEL_MARK if written[index] == LABEL_MARK else int(predicted[index])
        for index in range(len(typed))
    ]
    return apply_labels(typed, labels)


def window_bounds(length: int, window: int, overlap: int) -> list[tuple[int, int, int, int]]:
    """Split a long text into overlapping windows.

    Returns (start, end, commit_start, commit_end) per window: the model reads the whole
    window but only the committed part of it is used, so every predicted character had
    context on both sides.

    Every runtime - torch, onnxruntime in Python, onnxruntime in the browser - has to cut
    text the same way, or the same model gives different answers on long input. Keeping
    this in the torch-free module is what lets all of them share it.
    """
    if length <= window:
        return [(0, length, 0, length)]

    stride = window - overlap
    half = overlap // 2
    bounds: list[tuple[int, int, int, int]] = []

    for start in range(0, length, stride):
        end = min(start + window, length)
        commit_start = start if start == 0 else start + half
        commit_end = end if end == length else end - half
        bounds.append((start, end, commit_start, commit_end))
        if end == length:
            break

    return bounds


def foldable_positions(text: str) -> list[int]:
    """Indices where a decision is actually needed."""
    return [index for index, char in enumerate(text) if is_foldable(char)]


class TaggerRestorer:
    """Restores text with a character tagger.

    The predictor is injected, so the class can be tested without torch and can later be
    served from ONNX just as well as from a checkpoint.
    """

    name = "tagger"

    def __init__(
        self,
        predict: Callable[[Sequence[str]], list[list[int]]],
        config: TaggerConfig | None = None,
    ) -> None:
        self.predict = predict
        self.config = config or TaggerConfig()

    def restore(self, text: str) -> str:
        return self.restore_many([text])[0]

    def restore_many(self, texts: Sequence[str]) -> list[str]:
        """Restore a batch of texts; batching is what makes the model usable at scale."""
        if not texts:
            return []

        stripped = [strip_diacritics(text) for text in texts]
        predictions = self.predict(stripped)
        return [
            restore_with_labels(text, labels)
            for text, labels in zip(texts, predictions, strict=True)
        ]

    @classmethod
    def from_checkpoint(cls, path: Path, device: str | None = None) -> TaggerRestorer:
        """Load a trained checkpoint. Requires torch."""
        from schwa._tagger_model import load_predictor  # noqa: PLC0415

        predict, config = load_predictor(path, device=device)
        return cls(predict, config)


def config_to_dict(config: TaggerConfig) -> dict[str, float | int]:
    return asdict(config)


def config_from_dict(values: dict[str, float | int]) -> TaggerConfig:
    fields = {
        field: values[field] for field in TaggerConfig.__dataclass_fields__ if field in values
    }
    return TaggerConfig(**fields)  # type: ignore[arg-type]


__all__ += ["LABEL_KEEP", "config_from_dict", "config_to_dict", "foldable_positions"]
